source/client_channel_handler.c

```diff
@@ -602,8 +626,42 @@ static void s_request_timeout_task(struct aws_channel_task *task, void *arg, enu
             "%" PRIu16 ".",
             (void *)task,
             request->message_id);
-        /* If task cancelled, assume safe shutdown is in progress and signal hash_table destroy to clean up */
-        request->cancelled = true;
+        /*
+         * If the task is cancelled, assume safe shutdown is in progress.
+         * There are two distinct situations where requests end up with a cancelled task:
+         *   (1) The request has been completed but disconnect was begun before the task had a chance to run
+         *       normally.  We need a way to ensure that aws_memory_pool_release is called for this request.
+         *
+         *       We can do that by
+         *          (1) setting request->cancelled to true.  Then the hash table's value removal function will
+         *              ensure aws_memory_pool_release is called when the table is cleared during final disconnect.
+         *          (2) If we're uncomfortable (semantically) with a request that is both "completed" and "cancelled"
+         *              then we could alternatively call aws_memory_pool_release here.
```

Comment:
This comment will confuse future generations. Be more clear which of these things we're doing, or remove the alternative explanation of how we could do it.

Reply:
Yeah was probably better left as a review comment, but wanted people to read it and say which they preferred (if any).